### ISSUE TITLE
RFC: TransferQueue-Based Single-Pool Multi-Teacher Hidden-State OPD

### ISSUE THREAD

**Issue by uygnef** [verl-project/verl#6705]
## Summary

This RFC proposes a DeepSeek-V4-style OPD path in verl: use one shared teacher pool that switches teacher weights on demand, exports teacher hidden states through TransferQueue, and lets the student reconstruct full-vocabulary teacher logits for KL.

The main benefit is to support multi-teacher full-vocabulary OPD without keeping one resident serving pool per teacher or materializing large `[tokens, vocab]` teacher logits.

## Motivation

DeepSeek-V4 describes an efficient teacher scheduling design for full-vocabulary OPD: teacher weights are loaded on demand, teacher forward caches last-layer hidden states instead of full logits, training reconstructs full logits through the matching teacher head, and samples are ordered by teacher index so only one teacher head needs to be resident at a time.

Current verl MOPD supports selecting one teacher per trajectory and querying the corresponding teacher server for sampled-token or top-k logprobs. This works for top-k / sampled-token OPD, but scales poorly when:

- there are many specialist teachers;
- teacher models are large;
- we do not want every teacher to occupy a dedicated resident serving pool;
- full-vocabulary OPD / full-distribution KL is desired;
- we do not want to transmit or persist `[tokens, vocab]` teacher logits.

This RFC is a verl adaptation of that DeepSeek-V4-style scheduling pattern. It does not change teacher selection semantics. It changes teacher execution and teacher-signal data flow.

Current path:

```text
student trajectory
  -> selected teacher serving replica
  -> teacher sampled-token / top-k logprobs
  -> DataProto
  -> student loss
```

Proposed path:

```text
student trajectory
  -> selected teacher_index
  -> group by teacher_index
  -> shared teacher pool loads selected teacher weights
  -> teacher runs prefill/forward on prompt_ids + student_response_ids
  -> teacher writes final hidden states to TransferQueue
  -> TQ handles are scattered back to the original batch order
  -> student reconstructs full-vocab teacher logits and computes KL
```

## Related Work

- DeepSeek-V4, "Efficient Teacher Scheduling for Full-Vocabulary OPD"
  - Similarity: selected teacher index, grouped teacher execution, last-layer hidden-state cache, teacher-head-based full-logit reconstruction, and full-vocabulary KL.
  - Difference: DeepSeek-V4 describes a large-scale production training system with centralized distributed storage, asynchronous loading / unloading, and a custom TileLang KL kernel. This RFC scopes the first verl implementation to a single shared teacher pool, TransferQueue `SimpleStorage`, original batch-order handle scatter, and a non-fused chunked KL path.
- PR #6194, Nitrobrew OPD
  - Similarity: teacher hidden states and full-distribution KL.
  - Difference: this RFC focuses on a shared weight-swapped teacher pool and TQ-based hidden-state lifecycle.
  - If Nitrobrew hidden-state / KL primitives land first, this RFC can reuse its KL kernels, padding helpers, or teacher-head registration. The single-pool weight-swapping and TQ lifecycle design remain separate system work.
- PR #6243, vLLM PD disaggregated rollout with NIXL/Mooncake wiring
  - Similarity: disaggregated execution and large-tensor / cache transport.
  - Difference: this RFC applies to the OPD teacher path.
- PR #6373, MooncakeStoreConnector reset on weight update
  - Similarity: storage / cache correctness around weight versions.
  - Difference: the first version of this RFC does not depend on Mooncake.

## Goals

- Preserve existing MOPD semantics: each student-generated trajectory uses exactly one selected teacher.
- Make the DeepSeek-V4-style full-vocabulary OPD scheduling primitive explicit in verl.
- Reuse the existing teacher loop / resource manager / replica lifecycle as the base implementation path.
- Support one shared teacher pool serving multiple teachers.
- Group teacher work by `teacher_index` to reduce teacher weight switches.
- Support a teacher hidden-state signal.
- Store and transfer hidden states through TransferQueue instead of placing them directly in `DataProto`.
- Store only TQ handles / keys in the batch, then scatter them back to the original batch order.
- Reconstruct full-vocabulary teacher logits on the student side from hidden states and the matching teacher head.
- Support chunked full-vocabulary KL without keeping a long-lived `[tokens, vocab]` buffer.
- Use TQ `SimpleStorage` in the first version; do not depend on Mooncake or prefix hidden-state cache.
- Preserve the existing logprob / top-k OPD path as the default path.

## Non-Goals

- This is not best-of-N teacher selection.
- This does not require one live serving pool per teacher.
- The first version does not implement prefix hidden-state cache / prefix reuse.
- The first version does not depend on Mooncake / RDMA.
- The first version does not implement a fused KL kernel.
- This does not change rollout generation semantics.

## Scope

| Area | This RFC |
|---|---|
| Teacher selection | Reuse existing per-sample selected-teacher semantics |
| Teacher resource model | One shared teacher pool, one resident teacher at a time |
| Teacher backend | Reuse existing teacher loop / serving backend lifecycle; add backend-specific hidden-state extraction capability |
| Teacher signal | Add hidden-state signal; preserve logprob / top-k path |
| Hidden-state transport | TransferQueue |
| Batch payload | TQ handle / key only |
| Batch semantics | Teacher grouping affects execution order only; student update scatters handles back and aggregates in original batch order |
| KL path | Student side loads teacher head and computes full-vocab KL chunkwise |
| First PR | TQ SimpleStorage, no Mooncake, no prefix hidden-state cache |

## Architecture

```mermaid
flowchart LR
    subgraph StudentPool["Student Pool"]
        Rollout["Actor rollout workers<br/>generate trajectories"]
        Batch["Original rollout batch<br/>uid + batch_index + masks"]
        Update["Actor update workers<br/>fetch hiddens + compute KL"]
        HeadLoader["Teacher head loader<br/>sharded like student actor"]
    end

    subgraph TeacherPool["Shared Teacher Pool"]
        WeightSwapper["Weight switch<br/>one resident teacher"]
        Patch["Backend patch / hook<br/>capture final hidden states"]
        TeacherForward["Teacher prefill / forward only<br/>prompt_ids + response_ids"]
    end

    subgraph TQ["TransferQueue SimpleStorage"]
        HiddenStore[("Hidden-state records<br/>step-scoped, consume once")]
    end

    subgraph WeightStore["Teacher checkpoint / head storage"]
        Weights[("Teacher weights<br/>A / B / C")]
        Heads[("Teacher heads<br/>A / B / C")]
    end

    Rollout --> Batch
    Batch -->|"group by teacher_index<br/>execution only"| TeacherForward
    Weights -->|"load selected teacher"| WeightSwapper
    WeightSwapper --> TeacherForward
    Patch --> TeacherForward
    TeacherForward -->|"write hidden states"| HiddenStore
    HiddenStore -->|"return handle"| Batch
    Batch -->|"scatter handles back<br/>by batch_index"| Update
    Update -->|"get hidden states by handle"| HiddenStore
    Heads -->|"load matching teacher head"| HeadLoader
    HeadLoader --> Update
    Update -->|"chunked full-vocab KL<br/>original batch masks/order"| Update
```

### 1. Reuse Existing Teacher Assignment

This RFC does not change MOPD teacher assignment semantics. Each student-generated trajectory is still assigned to one selected teacher through the existing `distillation.teacher_key` / teacher routing logic.

The new path only normalizes the selected teacher into `teacher_index`, then groups teacher-signal computation by `teacher_index`.

If the same prompt samples multiple responses, each response is treated as an independent trajectory for teacher evaluation:

```text
1 generated trajectory -> 1 selected teacher evaluation
```

OPD itself does not require GRPO-style same-prompt multi-response group advantages. The first version targets pure OPD / distillation-loss training. If task rewards or GRPO-style advantages are later combined with this path, the sampler must preserve the grouping required by the algorithm.

### 2. Shared Weight-Swapped Teacher Pool

Add a teacher execution mode such as:

```yaml
distillation:
  enabled: true
  teacher_execution:
    mode: weight_swapped_hidden_state
    weight_cache:
      max_resident_teachers: 1
    hidden_state_transport:
      backend: transfer_queue
      storage_backend: simple
```

This is not a new serving backend. It is closer to the resource reuse pattern used by colocated rollout / sleep-wake / `update_weights`:

- the backend is still the existing vLLM / SGLang teacher serving backend;
- the teacher pool is still a resident set of Ray actors / serving workers;
- the backend is not destroyed or recreated between teachers;
- only the resident teacher weights change from teacher A to teacher B;
- the manager treats `teacher_models` as a teacher weight registry instead of splitting one resident pool per teacher.

Runtime flow:

1. Collect trajectories that require teacher evaluation.
2. Group them by `teacher_index`.
3. The shared teacher pool checks whether the current resident teacher matches the group.
4. If it does not match, drain the current group, then switch to the target teacher weights.
5. The teacher backend runs prefill / forward only on `prompt_ids + student_response_ids`; it does not decode or generate a new response.
6. The backend exports final hidden states aligned to the response loss positions.
7. Hidden states are written to TransferQueue.
8. TQ handles / keys are scattered back to the original batch order by `batch_index`.
9. Student update reads hidden states by handle and computes KL.

The first version can use blocking weight load. If weight switching becomes a bottleneck, async prefetch or a cross-microbatch teacher group queue can be added later.

The first version requires teacher compatibility at the serving layer:

- same architecture;
- same tokenizer / vocab;
- same hidden size;
- same TP / PP layout;
- same serving backend config.

Incompatible teacher configurations should fail during config validation.

### 3. Hidden-State Extraction

The current teacher serving path usually returns `prompt_logprobs`; it does not normally expose last hidden states. Therefore, this RFC does not assume the teacher backend is completely unchanged. It reuses the teacher loop management layer while adding backend-specific hidden-state extraction capability.

One implementation option is a torchspec-style backend patch / hook:

- patch or hook the teacher model forward path inside the backend;
- capture final decoder hidden states before the LM head;
- run on the concatenated student rollout sequence `prompt_ids + student_response_ids`;
- export hidden states aligned to response loss positions;
- write hidden states to TransferQueue;
- keep the existing serving lifecycle and request scheduling as much as possible.

The teacher does not perform rollout generation in this path. It only prefill/forwards the fixed sequence generated by the student.

The first version should support only one backend to avoid handling vLLM / SGLang / VLM differences at the same time.

The extraction path must guarantee:

- hidden states are attributed to the correct request / trajectory uid under concurrency;
- no full logits are materialized;
- only KL-needed token ranges are stored, preferably response / loss-mask positions;
- each record contains `teacher_key`, `teacher_index`, and `teacher_weight_version`.

### 4. TransferQueue-Based Hidden-State Store

Hidden states are not placed directly in `DataProto`. Large tensors go through TransferQueue; small metadata stays in the batch.

Teacher worker as producer:

```text
teacher compute hidden_states
  -> TQ put(hidden_state_record)
  -> return tq_handle
```

Student update / KL worker as consumer:

```text
read tq_handle from batch
  -> TQ get(hidden_state_record)
  -> reconstruct teacher logits
  -> compute KL
  -> TQ clear(hidden_state_record)
```

The first version uses step-scoped hidden-state records:

```text
produce once -> consume once -> clear
```

It does not reuse records across steps and does not implement prefix hidden-state cache.

Minimum record schema:

```text
hidden_state_record:
  uid
  batch_index
  teacher_key
  teacher_index
  teacher_weight_version
  sequence_hash
  seq_len
  token_range / loss_mask range
  hidden_shape
  hidden_dtype
  hidden_states
```

The consumer must validate `uid`, `teacher_index`, `teacher_weight_version`, and `sequence_hash` against batch metadata to avoid stale or mismatched hidden states.

### 5. Preserve Original Batch Order

Grouping by `teacher_index` affects teacher execution order only. It does not change the student training batch semantics.

The first version does not use ready-first consumption to form new training microbatches. TransferQueue is only used for large-tensor transport, not as the trainer scheduler.

Main path:

```text
rollout batch keeps original order
  -> each trajectory has uid / batch_index / selected_teacher_key
  -> group by teacher_index for teacher execution only
  -> shared teacher pool switches weights and computes hidden states group by group
  -> each hidden-state record is written to TQ and returns a handle
  -> handles are scattered back into the original batch by batch_index
  -> student update reads handles in original batch order
  -> KL is computed grouped by teacher_index
  -> KL loss is aggregated using original batch masks / token weights
```

This preserves:

- the rollout batch;
- the sample set of the optimizer step;
- loss mask / reward / advantage alignment;
- the batch semantics of other `DataProto` fields.

### 6. Teacher Head and Full-Vocab KL

The student side loads the teacher prediction head corresponding to `teacher_index`, then reconstructs teacher logits from teacher hidden states:

```text
teacher_logits = teacher_hidden_states @ teacher_lm_head
```

To control memory:

- process one teacher group at a time;
- keep at most one teacher head on device;
- shard the teacher head according to student actor parallelism;
- compute KL by token chunks / vocab chunks;
- free teacher logits after each chunk;
- never keep a long-lived full `[tokens, vocab]` tensor.

The first version requires teacher and student to share tokenizer and vocab mapping. Incompatible vocab should fail during config validation.

Teacher head sharding:

- FSDP: load / shard the teacher head according to actor FSDP layout.
- Megatron: follow actor tensor-parallel vocab sharding.
- Full-vocab KL uses the same distributed normalization pattern as the student actor's vocab-parallel loss path. Local vocab-shard logits are combined through all-reduce logsumexp / reductions when required.

### 7. Prefix Hidden-State Cache as Future Work

The first version does not implement prefix hidden-state cache. Prefix cache introduces additional correctness constraints:

- teacher weight version;
- sequence / prefix hash;
- partial token ranges;
- cache invalidation;
- cross-step lifecycle.

The first version should validate:

```text
single shared teacher pool
  + weight switch
  + hidden-state TQ transport
  + original batch-order handle scatter
  + student-side full-vocab KL
```

After that, future work can add:

- prefix hidden-state cache;
- TQ MooncakeStore / RayRDT backend;
- async teacher weight prefetch;

**Comment by uygnef** [comment]
@vermouth1992 Hi, could you take a look at this RFC

**Comment by moehanabi** [comment]
I think it's a great plan! Would you like to implement it?

**Comment by uygnef** [comment]
Yes, I’m planning to implement it. I opened this RFC first to get feedback from everyone on whether the direction fits the current roadmap, and whether there are any design or planning changes you would recommend before I start the implementation.
